py/desispec/scripts/stdstars.py

```diff
@@ -223,42 +222,30 @@ def main(args=None, comm=None) :
             frame = get_fiberbitmasked_frame(frame,bitmask='stdstars',ivar_framemask=True)
             frame_fibermap = frame.fibermap
 
-            # frame was filtered to just stdstars upon reading, so initial list of starindices is full range
-            frame_starindices = np.arange(len(frame.fibermap), dtype=int)
-
             #- Confirm that all fluxes have entries but trust targeting bits
             #- to get basic magnitude range correct
-            keep_legacy = np.ones(len(frame_starindices), dtype=bool)
-
+            keep_legacy = np.ones(len(frame_fibermap), dtype=bool)
```

Comment:
`keep_legacy`, `keep_gaia`, `n_legacy_std`, and `n_gaia_std` all seem to be redefined each loop here over both exposures (outer loop) and cameras (inner loop), but then used generically later in the code outside of any loops. Is there an assumption made here that allows that to be true?

Reply:
`scripts.stdstars` requires all input frames to be from the same spectrograph of the same tile, so if given correct inputs then this calculation should be redundant.  This block of code at the end of the loop confirms that the selected stars from each frame are consistent with the ones cached from the first frame, which is the core assumption made later in the code that needs to be checked before proceeding:
```python
            keep = keep_legacy | keep_gaia
            ...
            frame_starindices = frame_starindices[keep]

            if spectrograph is None :
                spectrograph = frame.spectrograph
                fibermap = frame_fibermap
                starindices=frame_starindices
                starfibers=np.asarray(fibermap["FIBER"][starindices])
            elif spectrograph != frame.spectrograph :
                log.error("incompatible spectrographs {} != {}".format(spectrograph,frame.spectrograph))
                raise ValueError("incompatible spectrographs {} != {}".format(spectrograph,frame.spectrograph))
            elif starindices.size != frame_starindices.size or np.sum(starindices!=frame_starindices)>0 :
                log.error("incompatible fibermap")
                raise ValueError("incompatible fibermap")
```
the most common user error would be providing frames from multiple spectrographs or multiple tiles, in which case simply checking FIBER and TARGETID values would be sufficient, but this loop adds additional checks on the consistency.  The calculation is ~0.005 sec per frame so the runtime is trivial.  There is in theory a pathological corner case where `keep_legacy` and `keep_gaia` could be different across frames while leaving `keep = keep_legacy | keep_gaia` the same, which this check would not catch.  It also wouldn't catch if the selected stars had "valid" but different photometry while still having the same fibers.

The structure of the loop pre-dates this PR and could be refactored for arguably better maintainability by separating out the concept of "keep" (on one frame) from checking the consistency of TARGETID, FIBER, FLUX* , GAIA_PHOT_* etc. across all frames.  But unless you think the current code is actively confusing or especially fragile, I suggest leaving it as is.

Reply:
Thannk you for the explanation. I now understand the flow of this and also agree that any inefficiency is trivial in the grand scheme of things.

Just for the record, the reference snippet above is the current `main` version of the code before your changes, which removed `keep`, `starindices`, and redefined  `starfibers`. I believe your points still stand, however, so I'll approve this once I test an example.